### PR TITLE
Initial refactoring of To_cmm

### PR DESCRIPTION
**Pull request by mshinwell**
There should be no changes here at all except refactoring.  Best read commit by commit.  The main things are:
- `To_cmm_helper` is no longer a dependency of `To_cmm_env` and `To_cmm_result`.  This means we can share more functions between different source files.
- The primitive simplification code has been moved to `To_cmm_primitive`.
- The code for dealing with the `Flambda_unit.t` itself is now in `To_cmm` and the old `To_cmm` was moved to `To_cmm_expr`.
- The recursive binding in `To_cmm_expr` has been significantly shrunk in size, which I think helps to understand the flow of what's going on, although it does mean that some of the helper functions are now further away.  Various functions have been simply inlined out; it was kind of hard to follow what was going on before, and this also reduces the risk of variable names diverging (which they had in some places).
- Some minor changes such as removing unnecessary `res` parameters in some places.

One of the main aims is to break the code down into smaller pieces so it's easier to check critical invariants, e.g. relating to linear use of the environment.

### CODE REVIEW

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml [hunk 1]
```diff
@@ -0,0 +1,825 @@
...
+       the argument to that continuation. Currently functions cannot return
+       unboxed values, so that argument must have a value kind. *)
+    ( wrap
+        (C.transl_switch_clambda Debuginfo.none (Vval Pgenval) scrutinee index
+           cases),
+      res )
+
+(* CR gbury: for the future, try and rearrange the generated cmm code to move
+   assignments closer to the variable definitions Or better: add traps to the
+   env to insert assignemnts after the variable definitions. *)
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_expr.ml:825]: I'm not sure whether this CR is still relevant. From what I recall, it was because originally the initialization code for the modules would end up looking like that:
```
let dummy_module_block = <initilization computation> in
let f1 = read_field dummy_module_block 1 in
let f2 = read_field dummy_module_block 2 in
...
write_field actual_module_block 1 f1;
write_field actual_module_block 2 f2;
...
```
Which would make values `f1`, `f2` and so on be live way longer than necessary.

However, I'm not sure we are still in that case, and things on the flambda2 simplifier side may have changed enough that we don't end up in that kind of situation often. So, we might need to check what happens and potentially remove this comment.

**Review by Gbury — CHANGES_REQUESTED** [review]
Mostly good but I left comments and I have a few overall remarks:
- outside of one or two shared/helper modules, code in `to_cmm` should never directly build cmm expressions using the datatype constructors; everything should go through smart constructors (as if the `cmm` types were private)
- we need a better and clearer distinction for what belongs in the `to_cmm_helper` module. what I propose is the following:
  + have all smart constructors for cmm be in either cmm_helpers or to_cmm_helpers (and thus if you prefer to not have a to_cmm_helper file, move all constructors to the cmm_helpers file) (and again ideally all of these should in my opinion be in the cmm module directly, but it's probably too late for that)
  + have simple/arg_list functions that consume flambda2 things and produce cmm be shared be in a `to_cmm_shared` file

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+  type cmm_term
+
+  val int : dbg:Debuginfo.t -> nativeint -> cmm_term
+
+  val simple :
+    To_cmm_env.t ->
+    Simple.t ->
+    [`Data of cmm_term list | `Var of Variable.t] * To_cmm_env.t * Ece.t
+
+  val infix_header : dbg:Debuginfo.t -> first_value_slot_offset:int -> cmm_term
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:59]: The name of the argument `first_value_slot_offset` is very confusing to me... What is actually given to the infix header is simply the offset of the closure inside the set of closures. The current name suggests (at least to me) a relation to the start of env, whereas there is none, so could you change the name ?

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+            get_whole_closure_symbol ~set_of_closures_symbol_ref
+          in
+          let env, updates =
+            C.make_update env dbg Word_val
+              ~symbol:(C.symbol set_of_closures_symbol)
+              v ~index:slot_offset ~prev_updates:updates
+          in
+          env, [P.int ~dbg 1n], updates
+      in
+      List.rev fields @ acc, slot_offset + 1, env, eff, updates
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:108]: We could use `List.rev_append`

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+    let term, env, eff = C.simple env simple in
+    `Data [term], env, eff
+
+  let infix_header ~dbg ~first_value_slot_offset =
+    C.alloc_infix_header first_value_slot_offset dbg
+
+  let symbol_from_linkage_name ~dbg linkage_name =
+    C.symbol_from_linkage_name ~dbg linkage_name
+
+  let define_global_symbol _ = []
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:186]: Maybe that should be an assert false ?

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+    let env, contents = C.simple_static env simple in
+    contents, env, Ece.pure
+
+  let infix_header ~dbg:_ ~first_value_slot_offset =
+    C.cint (C.infix_header first_value_slot_offset)
+
+  let symbol_from_linkage_name ~dbg:_ linkage_name =
+    C.symbol_address (Linkage_name.to_string linkage_name)
+
+  let define_global_symbol sym = List.rev (C.define_symbol ~global:true sym)
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:205]: I'd prefer that this function did not perform the `rev`, and instead the caller of this function (in the functor), performed the rev. This makes it much simpler to reason, and would allow to use `List.rev_append` which is more efficient (or at the very least, this function should be renamed with a `_rev` somewhere in its name).

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+      let acc =
+        if starting_offset > slot_offset
+        then
+          Misc.fatal_errorf "Starting offset %d is past slot offset %d"
+            starting_offset slot_offset
+        else if starting_offset = slot_offset
+        then acc
+        else
+          acc
+          @ List.init (slot_offset - starting_offset) (fun _ -> P.int ~dbg 1n)
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:159]: Shouldn't this concatenation be the other way ? I.e. put the dummy `1` elements at the front of the list, considering we build the reversed list of fields...

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml [hunk 7]
```diff
@@ -0,0 +1,431 @@
...
+  in
+  let block =
+    match l with
+    | _ :: _ ->
+      let header = C.cint (C.black_closure_header length) in
+      let sdef =
+        match !set_of_closures_symbol_ref with
+        | None -> []
+        | Some s ->
+          (* CR mshinwell: Is [global:false] correct? *)
```
Comment (Gbury) [inline middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:303]: Yes it is: basically, the symbol for the whole set of closures is generated by `to_cmm` when some updates are generated (since those updates will update the field of the set using the symbol for the whole set as a reference point), so it is only ever used locally by the updates expressions, which should be local (i.e. in the same module/object file than the set of closures).

File: ocaml/utils/target_system.mli [hunk 8]
```diff
@@ -9,6 +9,10 @@ type architecture =
 
 val architecture : unit -> architecture
 
+val is_64_bit : bool
+
+val is_32_bit : bool
```
Comment (Gbury) [inline ocaml/utils/target_system.mli:14]: Why do we need these additional functions when using `Arch.size_int` and other things already work ?

**Review by Gbury — CHANGES_REQUESTED** [review]
I left a lot of small comments, which have to be addressed.

Some notes:
- the formatting of `cmm_helpers` destroyed a lot of formatting in comments which have to be restored, and some parts of the code building some large cmm expressions became (even more) unreadable
- we will really need a second PR to harmonize `cmm_helpers.ml`: put the basic smart constructors for cmm expression at the top, and re-use them throughout the file, and try and decide on a common pattern for function signatures (i.e. labelled arguments or not, debuginfo as first or last argument, etc...).

File: middle_end/flambda2/to_cmm/to_cmm_env.mli [hunk 9]
```diff
@@ -43,10 +43,10 @@ val classify : Effects_and_coeffects.t -> kind
...
 type t
 
-(** [mk offsets k k_exn] creates a local environment for translating a flambda
-    expression, with return continuation [k], exception continuation [k_exn],
-    and which uses the given closures variables. *)
-val mk :
+(** [create offsets k k_exn] creates a local environment for translating a
+    flambda expression, with return continuation [k], exception continuation
+    [k_exn], and which uses the given closures variables. *)
+val create :
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_env.mli:49]: `create` is the usual name I think.

File: middle_end/flambda2/to_cmm/to_cmm_primitive.ml [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+(*                                                                        *)
+(*                        Guillaume Bury, OCamlPro                        *)
+(*                                                                        *)
+(*   Copyright 2019--2019 OCamlPro SAS                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_primitive.ml:13]: Yeah, this was all just copied over.

File: middle_end/flambda2/to_cmm/to_cmm_helper.ml [hunk 11]
```diff
@@ -645,3 +744,39 @@ let flush_cmmgen_state () =
     Misc.fatal_errorf
       "There shouldn't be any data item in cmmgen_state during flambda to cmm \
        translation"
+
+let make_update env kind ~symbol var ~index ~prev_updates =
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_helper.ml]: It's not just used in `To_cmm_static`, it's also used in `To_cmm_set_of_closures`, which is why it was moved.

File: middle_end/flambda2/to_cmm/to_cmm_expr.ml (hunk 1 above)
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_expr.ml:852]: Ah, this doesn't happen any more.  I'll remove the CR.

File: backend/cmm_helpers.ml [hunk 12]
```diff
@@ -120,473 +152,434 @@ let alloc_boxedintnat_header mode dbg =
...
+  | c1, Cconst_int (0, _) ->
+    Csequence (c1, raise_symbol dbg "caml_exn_Division_by_zero")
+  | c1, Cconst_int (1, _) -> c1
+  | Cconst_int (n1, _), Cconst_int (n2, _) -> Cconst_int (n1 / n2, dbg)
+  | c1, Cconst_int (n, _) when n <> min_int ->
+    let l = Misc.log2 n in
+    if n = 1 lsl l
+    then
+      (* Algorithm: t = shift-right-signed(c1, l - 1) t = shift-right(t, W - l)
+         t = c1 + t res = shift-right-signed(c1 + t, l) *)
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply backend/cmm_helpers.ml:473]: I'll deal with the ocamlformatting of `Cmm_helpers` in a separate PR to be merged before this one.

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+    startenv:int ->
+    Simple.t Value_slot.Map.t ->
+    Env.t ->
+    Ece.t ->
+    P.cmm_term list ->
+    Cmm.expression option ->
+    starting_offset:int ->
+    (int * Slot_offsets.layout_slot) list ->
+    P.cmm_term list * int * Env.t * Ece.t * Cmm.expression option
+end = struct
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:80]: I don't think blank lines in types are a good idea.  I'll look at the previous comment which might offer a way of dealing with this.

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml (hunk 4 above)
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:187]: Fixed

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml (hunk 7 above)
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:304]: OK

File: ocaml/utils/target_system.mli (hunk 8 above)
Comment (mshinwell) [reply ocaml/utils/target_system.mli:14]: Well there were two equivalents in `To_cmm_helpers` previously, and I don't think there is any other value that tells us the answer immediately.  We shouldn't be doing things like `when Arch.size_int = 4` which aren't robust.

**Comment by mshinwell** [comment]
Regarding a second PR to deal with moving functions around in `Cmm_helpers`, I think we should wait.  The desired situation is that Flambda 2 becomes the only middle-end, at which point a lot of the original Cmm generation code can be deleted.  That seems like the right time to make further improvements to the file.

File: middle_end/flambda2/to_cmm/to_cmm.ml [hunk 14]
```diff
@@ -29,1652 +27,91 @@ module R = To_cmm_result
...
-  let effs = Ece.all in
-  let fail_if_probe apply =
-    match Apply.probe_name apply with
-    | None -> ()
-    | Some _ ->
+(* Get constant definitions from Cmmgen_state. (The To_cmm translation uses
+   functions from Cmm_helpers which populate some mutable state in
+   Cmmgen_state.) *)
+
+let flush_cmm_helpers_state () =
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm.ml:39]: I don't think so, it's only used by this one place.  I pondered putting it in `Cmm_helpers` but eventually decided against.

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml (hunk 13 above)
[the remark this answers is not in the document]
Comment (Gbury) [reply middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:80]: Not necessarily in the type, but for instance, after the `struct`, to separate the `fill_slot` function from the functor signature

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml (hunk 2 above)
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:60]: Sorry, yes that isn't right.  Maybe it should be `function_slot_offset`?

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml (hunk 5 above)
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:206]: Fixed

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml (hunk 3 above)
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml]: Fixed

File: middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml [hunk 15]
```diff
@@ -0,0 +1,431 @@
...
+    Debuginfo.t ->
+    startenv:int ->
+    Simple.t Value_slot.Map.t ->
+    Env.t ->
+    Ece.t ->
+    P.cmm_term list ->
+    Cmm.expression option ->
+    starting_offset:int ->
+    (int * Slot_offsets.layout_slot) list ->
+    P.cmm_term list * int * Env.t * Ece.t * Cmm.expression option
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_set_of_closures.ml:79]: Fixed

File: middle_end/flambda2/to_cmm/to_cmm_env.ml [hunk 16]
```diff
@@ -222,7 +222,7 @@ let gen_variable v =
 
 let add_variable env v v' =
   let v'' = Backend_var.With_provenance.var v' in
-  let vars = Variable.Map.add v (To_cmm_helper.var v'') env.vars in
+  let vars = Variable.Map.add v (Cmm.Cvar v'') env.vars in
```
[the remark this answers is not in the document]
Comment (mshinwell) [reply middle_end/flambda2/to_cmm/to_cmm_env.ml]: Fixed (all occurrences of such things).

**Review by Gbury — APPROVED** [review]
That looks good, waiting for CI to be green.
